challenges/2021-11-16-data-grouping/solutions/javascript/justinmemphis/justinmemphis.js

```diff
@@ -0,0 +1,156 @@
+
+## Business Rules/Errata
+
+- You will be provided a file in the challenge setup with the data.
+- Use of external libraries, specifically data analysis libraries is encouraged.
+- The data output can be a dictionary or a dataframe (hint hint).
+- The data will be much larger than the example.
+- Stretch: plot the data in a meaningful way.
+
+## Examples
+
+Here is an example of the data file (tab delimited text):
+```
+HighSchool Year    ACT Score
+Sophomore          22
+Junior             28
+Freshman           21
+Freshman           19
+Junior             24
+Senior             25
+Senior             31
+Sophomore          31
+```
+Outputs of grouping (dataframe example):
+
+```
+HighSchool Year  Mean     Median
+Freshman         20       20
+Sophomore        26.5     26.5
+Junior           26       26
+Senior           28       28
+```
+
+###
+# outputs of:
+
+#return the number of scores greater than 25.
+3
+
+#return the number of sophomore scores greater than 25 and less than 30.
+0
+
+#return the mean of only the sophomore and junior scores.
+26.25
+
+## Tackling This Challenge
+
+1. Make sure you've got the required software on your machine: python 3.7+, pandas >= 1.0
+2. If you haven't already, fork the CodingDojo repository ([INSTRUCTIONS](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo)).
+4. The challenge will be live-coded in our weekly Tusday meetup in the `solution.py` file.
+6. The results of the live coding demo will be added to the GitHub repository as a mob solution. Please feel free to tackle this challenge in your own way/language and submit your solution to the repo using [this solutions guide in our wiki](https://github.com/codeconnector/CodingDojo/wiki#solutions).
+
+*/
+
+function mean(input) {
+	var sum = 0;
+	for (var i = 0; i < input.length; i++) {
+		sum += +input[i][1];
+	};
+	var result = sum / input.length;
+	return +result.toFixed(2);
+};
+
+function median(input) {
+	var numbers = [];
+	for (var i = 0; i < input.length; i++) {
+		numbers.push(+input[i][1]);
+	};
+
+	const sorted = Array.from(numbers).sort((a,b) => a - b);
+	const middle = Math.floor(sorted.length / 2);
+
+	if (sorted.length % 2 === 0) {
+		return (sorted[middle -1] + sorted[middle]) / 2;
+	}
+
+	return sorted[middle];
+};
+
+function higherThan25(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function twentyFiveToTwentyNine(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25 && +input[i][1] < 30) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function sortData(inputData, schoolYear, schoolYear2 = 'blah') {
+	let result = inputData.filter(element => element[0] == schoolYear || element[0] == schoolYear2);
+	return result;
+};
+
+function groupData() {
+	// type your solution here
+	fetch('http://justincarter.tech/HS_ACT_Data.txt')
```

Comment:
-1 for making me futz about with JavaScript! Nah, just kidding. I _do_ have suggestions for being able to run this locally, since apparently `fetch()` is a browser API (and is not stabilized on NodeJS, despite the numerous accounts on the internet saying it is). Most of our JS solutions assume you're running node locally. Instead of fetching from an external source, here's what your code would look like on Node pulling from a local file:

```javascript
const fs = require('fs');

function groupData() {
  const file = '../../../HS_ACT_Data.txt';
  const lines = fs.readFileSync(file).toString('UTF8');
  const data = lines.split('\r\n').map((line) => line.split('\t'));

  const freshmen = sortData(data, 'Freshman');
  const sophomores = sortData(data, 'Sophomore');
  const juniors = sortData(data, 'Junior');
  const seniors = sortData(data, 'Senior');
  const sophomoresAndJuniors = sortData(data, 'Sophomore', 'Junior');

  const bundle = {
    "Freshman Mean": round(mean(freshmen), 1),
    "Freshman Median": median(freshmen),
    "Sophomore Mean": round(mean(sophomores), 1),
    "Sophomore Median": median(sophomores),
    "Junior Mean": round(mean(juniors), 1),
    "Junior Median": median(juniors),
    "Senior Mean": round(mean(seniors), 1),
    "Senior Median": median(seniors),
    "Scores Over 25": higherThan25(data),
    "Sophomore Scores Between 25 and 30": twentyFiveToTwentyNine(sophomores),
    "Sophomore and Junior Mean": round(mean(sophomoresAndJuniors), 1)
  };

  return bundle;
};

module.exports = { groupData };
```

I'll say more about returning the `bundle` below.

Reply:
Oh, I wrote the `round()` function too. Here's that: 

```javascript
function round(value, precision) {
    var multiplier = Math.pow(10, precision || 0);
    return Math.round(value * multiplier) / multiplier;
}
```

Main reason is to help make the tests consistent and avoid floating point weirdness.

Reply:
Thank you!  I changed the language to add the fs module, add the bundle, and also add the rounding function.  I also removed all of my console.logs.
One change from what you have is I added the rounding function to the actual mean function, since we are using it every time we use mean.
See commit below.